vizro-core/tests/unit/vizro/models/_action/test_action.py

```diff
@@ -0,0 +1,205 @@
+            ),
+        ],
+    )
+    def test_inputs_outputs_valid(self, inputs, outputs, test_action_function):
+        action = Action(function=test_action_function, inputs=inputs, outputs=outputs)
+
+        assert action.inputs == inputs
+        assert action.outputs == outputs
+
+    @pytest.mark.parametrize(
+        "inputs",
+        [
+            None,
+            "string",
+            "component_property",
+            "compo-nent.property",
+            "component.property_1",
+        ],
+    )
+    def test_inputs_invalid(self, inputs, test_action_function):
+        with pytest.raises(Exception):
+            Action(function=test_action_function, inputs=inputs, outputs=[])
+
+    @pytest.mark.parametrize(
+        "outputs",
+        [
+            None,
+            "string",
+            "component_property",
+            "compo-nent.property",
+            "component.property_1",
+        ],
+    )
+    def test_outputs_invalid(self, outputs, test_action_function):
+        with pytest.raises(Exception):
+            Action(function=test_action_function, inputs=[], outputs=outputs)
+
+
+class TestActionBuild:
+    """Tests action build method."""
+
+    def test_build_no_returned_components_for_custom_action(self, test_action_function):
+        action = Action(function=test_action_function)
+        result = action.build()
+        expected = []
+        assert result == expected
+
+
+class TestActionPrivateMethods:
+    """Test action private methods."""
+
+    def test_get_callback_mapping_no_inputs_no_outputs(self, test_action_function):
+        action = Action(id="action_test", function=test_action_function)
+        callback_inputs, callback_outputs, action_components = action._get_callback_mapping()
+        assert callback_inputs == {
+            "trigger": dash.Input({"action_name": "action_test", "type": "action_trigger"}, "data")
+        }
+        assert callback_outputs == {"action_finished": dash.Output("action_finished", "data")}
+        assert action_components == []
+
+    @pytest.mark.parametrize(
+        "inputs, outputs",
+        [
+            (["component_1.property"], ["component_1.property"]),
+            (["component_1.property", "component_2.property"], ["component_1.property", "component_2.property"]),
+        ],
+    )
+    def test_get_callback_mapping_with_inputs_and_outputs(self, inputs, outputs, test_action_function):
+        action = Action(
+            function=test_action_function,
+            inputs=inputs,
+            outputs=outputs,
+        )
+        callback_inputs, callback_outputs, action_components = action._get_callback_mapping()
+        assert callback_inputs == {
+            **{
+                f'{input.split(".")[0]}_{input.split(".")[1]}': dash.State(input.split(".")[0], input.split(".")[1])
+                for input in inputs
+            },
+            "trigger": dash.Input({"action_name": action.id, "type": "action_trigger"}, "data"),
+        }
```

Comment:
Shall we move the expected result into fixtures? Would be more consistent with the rest of our unit tests. Same comment for code sections below

Reply:
Agreed, can we also remove some of the computation from the output? I would prefer to have a clear hard-coded dict as expected outcome:

1. This keeps tests pure, as in they don't mix function and result
2. it makes them far more readable as function documentation